fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java

```diff
@@ -94,16 +94,21 @@ private boolean isDataTypeValid(DataType originDataType, Expression expr) {
                                             (IntegralType) leftSlotEqualToRightSlot.child(1).getDataType())) {
                         return true;
                     }
+                } else if (leftSlotEqualToRightSlot.child(0).getDataType()
+                        .equals(leftSlotEqualToRightSlot.child(1).getDataType())) {
+                    return true;
                 }
                 return false;
             }
 
             private Expression replaceSlot(Expression expr, DataType originDataType) {
                 return expr.rewriteUp(e -> {
                     if (isDataTypeValid(originDataType, leftSlotEqualToRightSlot)) {
-                        if (ExpressionUtils.isTwoExpressionEqualWithCast(e, leftSlotEqualToRightSlot.child(0))) {
+                        if (ExpressionUtils.isTwoExpressionEqualWithCast(e, leftSlotEqualToRightSlot.child(0))
+                                && !(e instanceof Cast)) {
```

Comment:
why add `!(e instance Cast)` ? add some comment to explain it.
btw, `!(e instance Cast)` is a common predicate, should put at L106.
and `if (isDataTypeValid(originDataType, leftSlotEqualToRightSlot))` could put before return?
and, we do not use the second parameters at all in `isDataTypeValid` at all

Reply:
done

Reply:
because when rewriteup, it would do rewrite repeatly, assume we have a cast(a) = constant,and  a = b for example. When rewrite up, it would replace a to b first and then replace cast(b) -> a when go back to cast layer. isDatatypeValid parameter has been used.